### ISSUE TITLE
ClientHello 

### ISSUE THREAD

**Issue by 89z** [mitmproxy/mitmproxy#4877]
#### Problem Description

Is it possible to capture the TLS ClientHello for requests that go through MITM Proxy?

#### Proposal

Currently you can see all request headers and body, but is seems ClientHello is not captured. Binary output could be hex encoded, for example:

~~~
16030101fc010001f803036b27bf24169ff8443bdfbd0ad502eba49717c21df2166bd7cafa0b7e698fda3520c5ecabcfe5847cc0fbda97a1767863079eb57794869c07dea8a2ad2b9423fd320024130113031302c02bc02fcca9cca8c02cc030c00ac009c013c014009c009d002f0035000a0100018b0000001d001b000018636c69656e742e746c7366696e6765727072696e742e696f00170000ff01000100000a000e000c001d00170018001901000101000b00020100002300000010000e000c02683208687474702f312e310005000501000000000033006b0069001d00207a004cd760ae108579dcbd0eb81a77e02f4c82110dcb200e9742d6ec8e5dd25e001700410429fa3f3044cb41e2f5953348c540d88a6c7d29c48193d6519c1fcb402cd64f4b2a467c7e21c86dca36759d84c98f67c2927ea698cf4c3f1246c3209c0bf80f84002b00050403040303000d0018001604030503060308040805080604010501060102030201002d00020101001c00024001001500860000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
~~~

#### Alternatives

Currently I am using this site:

https://client.tlsfingerprint.io:8443

but that will only work for clients where you can manually make a request (such as a web browser, cURL etc). For clients where you cannot manually make a request, such as basically any APK, then it seems no way to capture the ClientHello.

#### Additional context

For servers that use TLS fingerprinting (`android.clients.google.com`, `www.amazon.com`), this would allow a user to capture the ClientHello of a known working client, and reuse that with future request on other clients.

**Comment by 89z** [comment]
Because of lack of response in other issue, I wrote my own proxy to capture TLS handshakes:

https://github.com/89z/parse/tree/master/cmd/proxy

**Comment by 89z** [comment]
I dont think this is fixed. Using this script:

~~~py
from mitmproxy import tls

class Counter:
   def tls_clienthello(self, data: tls.ClientHelloData):
      with open('hello.txt', 'w') as file:
         file.write(data.client_hello.raw_bytes.hex())

addons = [
    Counter()
]
~~~

I get this result:

~~~
03033b6b64b1bbcf681867e937b71d7645f786b12bf31b626e39842a3576c79d2b87000024c02bc02ccca9c02fc030cca8009e009fc009c00ac013c01400330039009c009d002f00350100006aff0100010000000022002000001d636c69656e7473657276696365732e676f6f676c65617069732e636f6d0017000000230000000d0016001406010603050105030401040303010303020102030010000b000908687474702f312e31000b00020100000a000400020017
~~~

A handshake starts with 0x16, which is missing from the output. Tools that parse handshakes expect this to be present [1]:

> The data should be the full tls record, including the record type/version/length header as well as the handshake type/length/version header https://tools.ietf.org/html/rfc5246#section-6.2 https://tools.ietf.org/html/rfc5246#section-7.4 

1. https://godocs.io/github.com/refraction-networking/utls#Fingerprinter.FingerprintClientHello

**Comment by mhils** [comment]
> A handshake starts with 0x16, which is missing from the output.

The ClientHello struct as defined in the RFC starts with the client_version/legacy_version bytes, which matches what mitmproxy returns now. If we want to be extra precise, everything around that is record layer / handshake wrapping, not ClientHello.

> Tools that parse handshakes expect this to be present [1]:

The same tool then seems to fail on fragmented TLS handshakes, so that argument is not entirely convincing. I do not blame them though, fragmented handshakes are hard to come by in practice. Theoretically speaking the already reassembled / unfragmented ClientHello we are providing right now should be easier to process. It's also nothing we've made up ourselves, see e.g. https://formats.kaitai.io/tls_client_hello/.

In either case, if it's helpful to API consumers to get a full record wrapper I'd be happy to accomodate. Given that this API is super new we can break things. Do you want to send a PR that:

1. Renames `.raw_bytes` to `._raw_bytes`
2. Adds a `raw_bytes(wrap_in_record: bool = True)` method, which depending on what is passed either returns `_raw_bytes` as-is or adds a single TLS record in front (i.e. 0x160303+len+0x01+len)?

We do need to make up a record layer protocol version for this (0x0303), but that isn't too much of a concern IMO. JA3 uses the ClientHello version anyhow, and we can point out that limitation in the docstring.

**Comment by 89z** [comment]
> The same tool then seems to fail on fragmented TLS handshakes, so that argument is not entirely convincing. I do not blame them though, fragmented handshakes are hard to come by in practice. Theoretically speaking the already reassembled / unfragmented ClientHello we are providing right now should be easier to process. It's also nothing we've made up ourselves, see e.g. https://formats.kaitai.io/tls_client_hello/.

For another data point, OpenSSL [1] returns the full TLS record as well:

~~~
<<< ??? [length 0005]
    16 03 01 02 00
<<< TLS 1.3, Handshake [length 0200], ClientHello
    01 00 01 fc 03 03 6c 72 13 f8 1b eb 21 77 3d 54
    f5 fd ed b9 20 71 75 77 b4 21 a4 ab 95 4e fc 4f
    fc 5d f9 91 cf cc 20 75 e6 87 f5 cb 0a 7a a4 14
    b7 9e 4c 62 43 16 6d 78 ca b0 1a 1a 3d 64 0a f8
    18 a9 ab f9 96 66 0e 00 3e 13 02 13 03 13 01 c0
    2c c0 30 00 9f cc a9 cc a8 cc aa c0 2b c0 2f 00
    9e c0 24 c0 28 00 6b c0 23 c0 27 00 67 c0 0a c0
    14 00 39 c0 09 c0 13 00 33 00 9d 00 9c 00 3d 00
    3c 00 35 00 2f 00 ff 01 00 01 75 00 00 00 0e 00
    0c 00 00 09 6c 6f 63 61 6c 68 6f 73 74 00 0b 00
    04 03 00 01 02 00 0a 00 0c 00 0a 00 1d 00 17 00
    1e 00 19 00 18 33 74 00 00 00 10 00 0e 00 0c 02
    68 32 08 68 74 74 70 2f 31 2e 31 00 16 00 00 00
    17 00 00 00 31 00 00 00 0d 00 30 00 2e 04 03 05
    03 06 03 08 07 08 08 08 09 08 0a 08 0b 08 04 08
    05 08 06 04 01 05 01 06 01 03 03 02 03 03 01 02
    01 03 02 02 02 04 02 05 02 06 02 00 2b 00 09 08
    03 04 03 03 03 02 03 01 00 2d 00 02 01 01 00 33
    00 26 00 24 00 1d 00 20 81 b6 5b 17 35 ec 43 59
    5e ba a5 4f 61 b0 0d ca 45 32 d3 4d 65 32 c2 87
    cd 11 f4 82 0f b2 a2 31 00 15 00 b4 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
~~~

In addition, another package I found also requires the full TLS record [2].

> In either case, if it's helpful to API consumers to get a full record wrapper I'd be happy to accomodate. Given that this API is super new we can break things. Do you want to send a PR that:

I cant do any PR at this time. If it means that issue is closed, so be it. However I think I have a case for the point about the full record, so it would be a shame if this feature gets so close to being useful without getting all the way there. Time will tell I suppose. Thank you for the fast and detailed reply.


1. https://unix.stackexchange.com/a/667815
2. https://godocs.io/github.com/dreadl0ck/tlsx#ClientHelloBasic.Unmarshal

**Comment by 89z** [comment]
I captured all the hellos from Android API 24 to 32 [1] using my proxy [2], so I guess I dont need this anymore. I think it would be useful in MITM proxy, but doesnt seem to be any interest in it.

1. https://github.com/89z/format/blob/v1.19.5/crypto/parse.go
2. https://github.com/89z/format/tree/v1.19.5/cmd/proxy

**Comment by mhils** [comment]
@89z: I'm not sure if my statement that I'd be happy to merge a PR wasn't clear enough or if I just misread your last comment as overly passive-aggressive, but I've implemented the changes I proposed in #5075 now.

**Comment by 89z** [comment]
No hard feeling, just in life people have different priorities. To me it is important to be able to capture the hellos, to others they might not even know or care what that is. I am not strong with Python, so I found my own solution with another language. I posted the issue in case anyone had a workaround or suggestion. It doesnt seem its an issue for others at this time. Thank you for the feedback and commit, you are a good person.
